### DISCUSSION TITLE
When navigating from foo?page=1 to foo?page=2, the foo/loading.js script is not triggered

### DISCUSSION THREAD

**Discussion by MelancholYA** [vercel/next.js#56657] [App Router]
I wrapped the element in suspense and add the search params string as key and it worked fine but only in development environment

**Comment by SidKH** [comment]
The `loading.js` will not get triggered in this case since it's the same page.

Instead, as you noted, you need to wrap **the element that does the fetching** in Suspense. 
It should work in dev and prod. 

```jsx
import { Suspense } from "react";

async function ComponentThatDoesTheFetching() {
  await fetch(...)

  return ...;
}

export default function Page({ params, searchParams }) {
  return (
    <div>
      <Suspense
        key={searchParams.myParam}
        fallback="Loading..."
      >
        <ComponentThatDoesTheFetching />
      </Suspense>
      ...
    </div>
  );
}
```

Also, consider providing an inline loading indicator instead of re-triggering Suspense. 
https://twitter.com/asidorenko_/status/1700102583208034746

**Reply by MelancholYA** [reply]
in my case the loading only worked in ther dev env , here's the behavior : https://www.loom.com/share/54961dffd5a740dfb0f702c0eada2462 


the suspense :

 ```<div className=" px-1 lg:px-5 shrink-0 lg:w-4/5">
          <Suspense key={JSON.stringify(searchParams)}>
            <Products searchParams={searchParams} />
          </Suspense>
        </div> ```

products component : 

 ```const Products: FC<Props> = async ({ searchParams }) => {
  const params = strToQuery(searchParams);
  const res = await axios.get("product", { params });
  const { items }: responseType = res.data;
  return !items || !items.length ? (
    <NoData />
  ) : (
    <div className="flex flex-wrap  items-center justify-between">
      {items.map((item) => (
        <ProductCard
          key={`tshirt-key-${item.id}`}
          image={item.thumbnail}
          name={item.name}
          price={`${item.price} DZD`}
          slug={item.slug}
        />
      ))}
    </div>
  );
}; ```

**Comment by SidKH** [comment]
```
<Suspense key={JSON.stringify(searchParams)}>
  <Products searchParams={searchParams} />
</Suspense>
```

Does your suspense have a `fallback` prop?

**Reply by MelancholYA** [reply]
i'm using nextjs's dynamic so the fallback is included in the import , i tried Lazy before but they all act the same

```
const Products = dynamic(() => import("@/components/store/Products"), {
  loading: () => <Loader />,
});
```

**Reply by SidKH** [reply]
Why do you need to dynamically import `Products`? It seems to be a server component.

**Reply by MelancholYA** [reply]
will the suspense work on a normal import ?
since it's an async component  and it will be refetched when query changes i thought it's better to dynamically import it

**Reply by SidKH** [reply]
Yep, you don't need dynamic imports to use `Suspense` in server components.

Just import your server component directly and include a fallback to your Supsense.
  
```jsx
import Products from '@/components/store/Products';


function Page() {
  ...
  return (
   ....
    <Suspense fallback={<Loader />} key={JSON.stringify(searchParams)}>
      <Products searchParams={searchParams} />
    </Suspense>
    ....
  )
}
```

**Reply by MelancholYA** [reply]
still not working , the problem only happens after deployment

**Reply by SidKH** [reply]
The purpose of dynamic imports is to defer the loading of the **client components javascript** to reduce initial bundle size.
  
You don't need this for server components since server components run entirely on the server and only ship the results of their execution to the client.

The `Suspense` on the server will get re-triggered every time the server component re-renders. And server components will re-render during hard or soft navigation (when you push new query param - `Products` re-renders and re-fetches data)

**Reply by MelancholYA** [reply]
when serving the build version locally  it works fine but in the server it doesnt , maybe it has something to do with my nginx config or something ?
